R-package/src/lightgbm_R.cpp

```diff
@@ -616,31 +606,33 @@ SEXP LGBM_BoosterSaveModel_R(LGBM_SE handle,
 
 SEXP LGBM_BoosterSaveModelToString_R(LGBM_SE handle,
   SEXP num_iteration,
-  SEXP feature_importance_type,
-  SEXP buffer_len,
-  SEXP actual_len,
-  LGBM_SE out_str) {
+  SEXP feature_importance_type) {
+  SEXP model_str;
   R_API_BEGIN();
   int64_t out_len = 0;
-  int64_t buf_len = static_cast<int64_t>(Rf_asInteger(buffer_len));
+  int64_t buf_len = 1024 * 1024;
```

Comment:
What if `1024 * 1024` is not enough to save some big model? With removed "try with default len, repeat with actual if not enough" (`if (act_len > buf_len)`) this now looks like a regression compared to the current fully correct implementation.

C API docs says:
> **buffer_len** – String buffer length, if `buffer_len < out_len`, you should re-allocate buffer

Reply:
OHHHH I see now, thank you for that explanation. I misunderstood the purpose of the code on the R side that was calling this function twice.

Ok yes you're right, that work needs to be done here. Will update it.

Reply:
Alright, I've made this changes in recent commits.

I'm really glad you point this out, because it also made me realize an opportunity to allow larger feature names! Right now, the code in `LGBM_DatasetGetFeatureNames_R` on `master` will not allow any feature names longer than 256 characters.

https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/R-package/src/lightgbm_R.cpp#L163

https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/R-package/src/lightgbm_R.cpp#L179

The code below throws an error on `{lightgbm}` 3.2.1, but works as of this branch.

```r
library(lightgbm)

feature_names <- names(iris)
long_name <- paste0(rep("a", 1000L), collapse = "")
feature_names[1L] <- long_name
names(iris) <- feature_names
# check that feature name survived the trip from R to C++ and back
dtrain <- lgb.Dataset(
    data = as.matrix(iris[, -5L])
    , label = as.numeric(iris$Species) - 1L
)
dtrain$construct()
col_names <- dtrain$get_colnames()

# Error in lgb.call(fun_name = fun_name, ret = buf, ..., buf_len, act_len) : 
#  [LightGBM] [Fatal] Check failed: (reserved_string_size) >= (required_string_size) at lightgbm_R.cpp, line 177 .
```

But it should be possible to! Based on https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/include/LightGBM/c_api.h#L295.

So I've updated the calls to `LGBM_DatasetGetFeatureNames_R` and `LGBM_BoosterGetEvalNames_R` to retry with a larger buffer on long names.

<hr>

I've added tests to this PR to check that this is working as expected.

* `LGBM_DatasetGetFeatureNames_R()`: `test_dataset.R`
* `LGBM_BoosterGetEvalNames_R()`:  I could not find a way to generate a large string value for this, but I might misunderstand how `LGBM_BoosterGetEvalNames` works. Opened #4264 with a question.
* `LGBM_BoosterSaveModelToString_R()`: `test_lgb.Booster.R`
* `LGBM_BoosterDumpModel_R()`: `test_lgb.Booster.R`

Reply:
> So I've updated the calls to `LGBM_DatasetGetFeatureNames_R` and `LGBM_BoosterGetEvalNames_R` to retry with a larger buffer on long names.

Great!
I think we can do the same for Python wrapper.
https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/python-package/lightgbm/basic.py#L1896-L1900
https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/python-package/lightgbm/basic.py#L3273-L3277
https://github.com/microsoft/LightGBM/blob/f83180883a51729af9562ccac7b450c6da79a04a/python-package/lightgbm/basic.py#L3472-L3476

Reply:
> I think we can do the same for Python wrapper.

Sure! But I think it should be a separate PR. I'd really like to focus on finishing #3016 soon.